book/src/qobject/qobject_struct.md

```diff
@@ -5,42 +5,140 @@ SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+Invokables are functions that are registered with the Qt meta-object system using [`Q_INVOKABLE`](https://doc.qt.io/qt-6/qobject.html#Q_INVOKABLE).
+This allows them to be called from QML and JavaScript (running in a [QJSEngine](https://doc.qt.io/qt-6/qjsengine.html)), in addition to C++ and Rust.
 
-To mark a method as invokable simply add the `#[qinvokable]` attribute to the Rust method. This then causes `Q_INVOKABLE` to be set on the C++ definition of the method, allowing QML to call the invokable.
+CXX-Qt allows you to define invokables using Rust code.
+This way you can easily add a Rust-powered backend to your QML frontend.
 
-Note to access properties on the C++ object use [Cpp Object](./cpp_object.md).
+Invokables, by definition, must be defined on a C++ class however.
+This is where the QObject subclass generated by `#[cxx_qt::qobject]` comes into play.
+For details on this, see the [`qobject::T` page](./cpp_object.md).
 
-## Properties
+The important part for invokables is that they need to be implemented on the `qobject::T`, not `T`.
+Therefore they have access to both C++ and Rust methods. CXX-Qt adds wrapper code around your invokables to automatically convert between the [C++ and Rust types](../concepts/types.md).
 
-Fields within the `#[cxx_qt::qobject]` marked struct can be tagged with the `#[qproperty]`, this declares which fields should be exposed as `Q_PROPERTY`s on the QObject.
+To mark a method as invokable simply add the `#[qinvokable]` attribute to the Rust method. This then causes the method to be exposed on the C++ class.
+`Q_INVOKABLE` will also be added to the definition of the method, allowing QML to call the invokable.
 
-```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/rust/src/struct_properties.rs:book_macro_code}}
+``` rust,ignore,noplayground
+{{#include ../../../examples/qml_features/rust/src/invokables.rs:book_impl_qobject}}
 ```
 
-If you want to provide default values for your QObject, then instead of deriving implement the `Default` trait for the struct.
+Note that an Invokable may only use `self: Pin<&mut Self>` or `&self` as self types.
+It is not possible to have a `self`, or `&mut self` invokable, as that may move around the QObject, which C++ can't handle.
+Furthermore, invokables are restricted to only use types that are compatible with CXX.
+
+It is also possible to define methods in the `impl qobject::T` block that are *not* marked as `#[qinvokable]`.
+These methods won't be available from C++ or QML.
+But as they still use the QObject as the self type they can still access the QObject features like emitting signals, changing properties, etc.
+However, they are still normal Rust functions.
+Therefore, they aren't restricted to CXX-compatible types.
+
+[Full example](https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/invokables.rs)
 
 ## Private Methods and Fields
```

Comment:
should you mention in this section about the `field` / `field_mut` / `set_field` ? and how `field_mut` is safe as it's not a qproperty ?

Reply:
I've moved all information about the non-property fields into the `Private Methods and Fields` section and linked to it from the properties section.